src/systems/tips.js

```diff
@@ -0,0 +1,274 @@
+import { sets } from "./userinput/sets";
+import { paths } from "./userinput/paths";
+import { detectInHMD } from "../utils/vr-caps-detect";
+
+// The output of this system is activeTips which shows, if any, the tips to show at the top
+// and bottom of the screen. There are named tips (eg locomotion) that each have validators.
+//
+// Each frame we run all the non-finished validators and take the first tip (by order)
+// which is VALID. Any tip that returns FINISHED will no longer be considered without
+// a local storage reset.
+
+// Validators can return these values:
+//
+// INVALID - Tip is not valid to show
+// VALID - Tip is valid to show
+// FINISH - Tip should always be assumed INVALID going forward
+//
+//
+// Tips have "scope", which is a certain context within which the validators are run (right now,
+// just the top of bottom of the screen.) They are also platform-local.
+const INVALID = 0;
+const VALID = 1;
+const FINISH = 2;
+
+const LOCAL_STORAGE_KEY = "__hubs_finished_tips";
+
+const TIPS = {
+  desktop: {
+    top: ["mirror_mode", "pen_mode", "video_share_mode", "mute_mode"],
+    bottom: [
+      "look",
+      "locomotion",
+      "turning",
+      "spawn_menu",
+      "object_grab",
+      "object_zoom",
+      "object_scale",
+      "freeze_gesture",
+      "menu_hover",
+      "object_pin",
+      "invite",
+      "pen_color",
+      "pen_size"
+    ]
+  },
+  mobile: {
+    top: ["pen_mode", "video_share_mode", "freeze_mode", "mute_mode"],
+    bottom: ["look", "locomotion", "spawn_menu", "object_grab", "freeze_gesture", "object_pin", "invite"]
+  },
+  standalone: { top: [], bottom: [] }
+};
+
+let localStorageCache = null;
+let finishedScopes = {}; // Optimization, lets system skip scopes altogether once finished.
+
+const isMobile = AFRAME.utils.device.isMobile();
+
+const tipPlatform = () => {
+  if (detectInHMD()) return "standalone";
+  return isMobile ? "mobile" : "desktop";
+};
+
+const platformTips = () => {
```

Comment:
Since platform can not change, instead of doing this I might pull the current platforms tips out into another const. Then, remove all the finished ones. And in `markTipFinished` remove it from the array. That way you never bother looping though tips in the (very common) case you have completed them. And in the case where we determine at startup that you have no tips to complete we can just not bother doing any work at all.

Reply:
The first part makes sense, but mutating the array could get ugly because of the "Show Tips" reset button

Reply:
Ah makes sense, missed that feature